externals: nitro is not picking the latest version

### Environment

------------------------------
- Operating System: `Windows_NT`
- Node Version:     `v16.11.1`
- Nuxt Version:     `3.0.0-rc.10`
- Package Manager:  `npm@8.0.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------
Nitro 0.5.3

### Reproduction

https://github.com/huang-julien/nuxt-nitro-package-external-yarn yarn package manager 
https://github.com/huang-julien/nuxt-nitro-package-external-npm npm pacakge manager

### Describe the bug

Hi 👋 i ran into this issue when using the nuxt-runtime-compiler-module and updating the runtime-compiler PR on nuxt.

Basically it seems that externals plugin should be picking the last version of estree-walker (3.0.1). However the package.json sent to .output is the json file from 2.0.2.

This is reproductible with npm but not yarn because yarn install estree-walker@2.0.2 to node_modules while npm install estree-walker@3.0.1 to node_modules and estree-walker@2.0.2 to node_modules/@vue/compiler-core.

this is from 3.0.1
```json
{
  "name": "estree-walker",
  "description": "Traverse an ESTree-compliant AST",
  "version": "3.0.1",
  "private": false,
  "author": "Rich Harris",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/Rich-Harris/estree-walker"
  },
  "type": "module",
  "module": "./src/index.js",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": "./src/index.js"
    }
  },
  "types": "types/index.d.ts",
  "scripts": {
    "prepublishOnly": "tsc && npm test",
    "test": "uvu test"
  },
  "devDependencies": {
    "@types/estree": "0.0.42",
    "typescript": "^3.7.5",
    "uvu": "^0.5.1"
  },
  "files": [
    "src",
    "types",
    "README.md"
  ]
}

```

This is from the 2.0.2 and the one that is being send to .output (on npm and yarn)

```json
{
  "name": "estree-walker",
  "description": "Traverse an ESTree-compliant AST",
  "version": "2.0.2",
  "private": false,
  "author": "Rich Harris",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/Rich-Harris/estree-walker"
  },
  "type": "commonjs",
  "main": "./dist/umd/estree-walker.js",
  "module": "./dist/esm/estree-walker.js",
  "exports": {
    "require": "./dist/umd/estree-walker.js",
    "import": "./dist/esm/estree-walker.js"
  },
  "types": "types/index.d.ts",
  "scripts": {
    "prepublishOnly": "npm run build && npm test",
    "build": "tsc && rollup -c",
    "test": "uvu test"
  },
  "devDependencies": {
    "@types/estree": "0.0.42",
    "rollup": "^2.10.9",
    "typescript": "^3.7.5",
    "uvu": "^0.5.1"
  },
  "files": [
    "src",
    "dist",
    "types",
    "README.md"
  ]
}

````

The problem with npm is that the files from estree-walker@3.0.1 and the package.json from 2.0.2 is sent but the exports defined in the package json does not exist.

I also tried to force include it using traceInclude without success

maybe this issue shoud be renamed ? i couldn't find a proper title

### Additional context

related https://github.com/nuxt/framework/pull/4762
related https://github.com/huang-julien/nuxt-runtime-compiler/issues/19

### Logs

```sh
Log during build 

 WARN  Multiple major versions of package estree-walker are being externalized. Picking latest version.                                                                             00:25:31  
  - /GIT/nuxt-version-diff-npm/node_modules/@vue/compiler-core/node_modules/estree-walker@2.0.2
  - /GIT/nuxt-version-diff-npm/node_modules/estree-walker@3.0.1

Log when going to /
[nuxt] [request error] [unhandled] [500] Cannot find module 'D:\GIT\nuxt-version-diff-npm\.output\server\node_modules\estree-walker\dist\esm\estree-walker.js' imported from D:\GIT\nuxt-version-diff-npm\.output\server\chunks\renderer.mjs
```

--- a/src/options.ts
+++ b/src/options.ts
@@ -155,7 +155,7 @@ export async function loadOptions (configOverrides: NitroConfig = {}): Promise<N
   options.nodeModulesDirs.push(resolve(options.workspaceDir, 'node_modules'))
   options.nodeModulesDirs.push(resolve(options.rootDir, 'node_modules'))
   options.nodeModulesDirs.push(resolve(pkgDir, 'node_modules'))
-  options.nodeModulesDirs = Array.from(new Set(options.nodeModulesDirs))
+  options.nodeModulesDirs = Array.from(new Set(options.nodeModulesDirs.map(dir => resolve(options.rootDir, dir))))
 
   if (!options.scanDirs.length) {
     options.scanDirs = [options.srcDir]

--- a/src/rollup/plugins/externals.ts
+++ b/src/rollup/plugins/externals.ts
@@ -196,11 +196,7 @@ export function externals (opts: NodeExternalsOptions): Plugin {
           }
 
           // Exclude older version files
-          if (isNewer) {
-            ignoreDirs.push(existingPkgDir)
-          } else {
-            ignoreDirs.push(pkgDir)
-          }
+          ignoreDirs.push(isNewer ? existingPkgDir : pkgDir)
         } else {
           // Add to traced packages
           tracedPackages.set(pkgName, pkgDir)

--- a/src/rollup/plugins/externals.ts
+++ b/src/rollup/plugins/externals.ts
@@ -174,10 +174,10 @@ export function externals (opts: NodeExternalsOptions): Plugin {
         if (existingPkgDir && existingPkgDir !== pkgDir) {
           const v1 = await getPackageJson(existingPkgDir).then(r => r.version)
           const v2 = await getPackageJson(pkgDir).then(r => r.version)
-          if (semver.gte(v1, v2)) {
-            // Existing record is newer or same. Just skip.
-            continue
-          }
+          // if (semver.gte(v1, v2)) {
+          // Existing record is newer or same. Just skip.
+          // continue
+          // }
           if (semver.major(v1) !== semver.major(v2)) {
             console.warn(`Multiple major versions of package ${pkgName} are being externalized. Picking latest version.\n` + [
               existingPkgDir + '@' + v1,

--- a/src/rollup/plugins/externals.ts
+++ b/src/rollup/plugins/externals.ts
@@ -173,7 +173,7 @@ export function externals (opts: NodeExternalsOptions): Plugin {
         if (!pkgName) {
           continue
         }
-        const pkgDir = resolve(baseDir, pkgName)
+        let pkgDir = resolve(baseDir, pkgName)
 
         // Check for duplicate versions
         const existingPkgDir = tracedPackages.get(pkgName)
@@ -196,11 +196,16 @@ export function externals (opts: NodeExternalsOptions): Plugin {
           }
 
           // Exclude older version files
-          ignoreDirs.push(isNewer ? existingPkgDir : pkgDir)
-        } else {
-          // Add to traced packages
-          tracedPackages.set(pkgName, pkgDir)
+          if (isNewer) {
+            ignoreDirs.push(existingPkgDir)
+          } else {
+            ignoreDirs.push(pkgDir)
+            pkgDir = existingPkgDir // Update for tracedPackages
+          }
         }
+
+        // Add to traced packages
+        tracedPackages.set(pkgName, pkgDir)
       }
 
       // Filter out files from ignored packages and dedup

--- a/src/rollup/plugins/externals.ts
+++ b/src/rollup/plugins/externals.ts
@@ -1,5 +1,6 @@
 import { existsSync, promises as fsp } from 'fs'
 import { resolve, dirname, normalize, join, isAbsolute } from 'pathe'
+import consola from 'consola'
 import { nodeFileTrace, NodeFileTraceOptions } from '@vercel/nft'
 import type { Plugin } from 'rollup'
 import { resolvePath, isValidNodeImport, normalizeid } from 'mlly'
@@ -145,9 +146,12 @@ export function externals (opts: NodeExternalsOptions): Plugin {
       }
 
       // Trace files
-      const tracedFiles = await nodeFileTrace(Array.from(trackedExternals), opts.traceOptions)
+      let tracedFiles = await nodeFileTrace(Array.from(trackedExternals), opts.traceOptions)
         .then(r => Array.from(r.fileList).map(f => resolve(opts.traceOptions.base, f)))
         .then(r => r.filter(file => file.includes('node_modules')))
+
+      // Resolve symlinks
+      tracedFiles = await Promise.all(tracedFiles.map(file => fsp.realpath(file)))
 
       // Read package.json with cache
       const packageJSONCache = new Map() // pkgDir => contents
@@ -162,6 +166,8 @@ export function externals (opts: NodeExternalsOptions): Plugin {
 
       // Keep track of npm packages
       const tracedPackages = new Map() // name => pkgDir
+      const ignoreDirs = []
+      const ignoreWarns = new Set()
       for (const file of tracedFiles) {
         const { baseDir, pkgName } = parseNodeModulePath(file)
         if (!pkgName) {
@@ -174,20 +180,35 @@ export function externals (opts: NodeExternalsOptions): Plugin {
         if (existingPkgDir && existingPkgDir !== pkgDir) {
           const v1 = await getPackageJson(existingPkgDir).then(r => r.version)
           const v2 = await getPackageJson(pkgDir).then(r => r.version)
-          // if (semver.gte(v1, v2)) {
-          // Existing record is newer or same. Just skip.
-          // continue
-          // }
-          if (semver.major(v1) !== semver.major(v2)) {
-            console.warn(`Multiple major versions of package ${pkgName} are being externalized. Picking latest version.\n` + [
-              existingPkgDir + '@' + v1,
-              pkgDir + '@' + v2
-            ].map(p => '  - ' + p).join('\n'))
+          const isNewer = semver.gt(v2, v1)
+
+          // Warn about major version differences
+          const getMajor = v => v.split('.').filter(s => s !== '0')[0]
+          if (getMajor(v1) !== getMajor(v2)) {
+            const warn = `Multiple major versions of package \`${pkgName}\` are being externalized. Picking latest version:\n\n` + [
+              `  ${isNewer ? '-' : '+'} ` + existingPkgDir + '@' + v1,
+              `  ${isNewer ? '+' : '-'} ` + pkgDir + '@' + v2
+            ].join('\n')
+            if (!ignoreWarns.has(warn)) {
+              consola.warn(warn)
+              ignoreWarns.add(warn)
+            }
           }
+
+          // Exclude older version files
+          if (isNewer) {
+            ignoreDirs.push(existingPkgDir)
+          } else {
+            ignoreDirs.push(pkgDir)
+          }
+        } else {
+          // Add to traced packages
+          tracedPackages.set(pkgName, pkgDir)
         }
-        // Add to traced packages
-        tracedPackages.set(pkgName, pkgDir)
       }
+
+      // Filter out files from ignored packages
+      tracedFiles = tracedFiles.filter(f => !ignoreDirs.some(d => f.startsWith(d)))
 
       // Ensure all package.json files are traced
       for (const pkgDir of tracedPackages.values()) {

--- a/src/rollup/plugins/externals.ts
+++ b/src/rollup/plugins/externals.ts
@@ -207,8 +207,9 @@ export function externals (opts: NodeExternalsOptions): Plugin {
         }
       }
 
-      // Filter out files from ignored packages
+      // Filter out files from ignored packages and dedup
       tracedFiles = tracedFiles.filter(f => !ignoreDirs.some(d => f.startsWith(d)))
+      tracedFiles = Array.from(new Set(tracedFiles))
 
       // Ensure all package.json files are traced
       for (const pkgDir of tracedPackages.values()) {
@@ -226,14 +227,9 @@ export function externals (opts: NodeExternalsOptions): Plugin {
         await fsp.mkdir(dirname(dst), { recursive: true })
         await fsp.copyFile(src, dst)
       }
-      if (process.platform === 'win32') {
-        // Workaround for EBUSY on windows (#424)
-        for (const file of tracedFiles) {
-          await writeFile(file)
-        }
-      } else {
-        await Promise.all(tracedFiles.map(writeFile))
-      }
+
+      // Write traced files
+      await Promise.all(tracedFiles.map(writeFile))
 
       // Write an informative package.json
       await fsp.writeFile(resolve(opts.outDir, 'package.json'), JSON.stringify({

--- a/test/fixture/_/.gitignore
+++ b/test/fixture/_/.gitignore
@@ -1,0 +1,1 @@
+!node_modules

--- a/test/fixture/nitro.config.ts
+++ b/test/fixture/nitro.config.ts
@@ -21,6 +21,9 @@ export default defineNitroConfig({
       dir: 'public/build'
     }
   ],
+  nodeModulesDirs: [
+    './_/node_modules'
+  ],
   prerender: {
     crawlLinks: true,
     ignore: [

--- a/test/fixture/routes/modules.ts
+++ b/test/fixture/routes/modules.ts
@@ -1,5 +1,8 @@
+// @ts-ignore
 import depA from 'nitro-dep-a'
+// @ts-ignore
 import depB from 'nitro-dep-b'
+// @ts-ignore
 import depLib from 'nitro-lib'
 
 export default defineEventHandler(() => {

--- a/test/fixture/routes/modules.ts
+++ b/test/fixture/routes/modules.ts
@@ -1,0 +1,11 @@
+import depA from 'nitro-dep-a'
+import depB from 'nitro-dep-b'
+import depLib from 'nitro-lib'
+
+export default defineEventHandler(() => {
+  return {
+    depA,
+    depB,
+    depLib
+  }
+})

--- a/test/fixture/types.ts
+++ b/test/fixture/types.ts
@@ -15,6 +15,7 @@ describe('API routes', () => {
       | string | number | TestResponse
       | { testFile: string, hasEnv: boolean }
       | { internalApiKey: string }
+      | { [key: string]: any }
       >>() // middleware
     expectTypeOf($fetch('/api/unknown')).toMatchTypeOf<Promise<unknown>>()
     expectTypeOf($fetch<TestResponse>('/test')).toMatchTypeOf<Promise<TestResponse>>()

--- a/test/tests.ts
+++ b/test/tests.ts
@@ -134,5 +134,10 @@ export function testNitro (ctx: Context, getHandler: () => TestHandler | Promise
       const { status } = await callHandler({ url: '/build/non-file' })
       expect(status).toBe(404)
     })
+
+    it('resolve module version conflicts', async () => {
+      const { data } = await callHandler({ url: '/modules' })
+      expect(data).toMatchObject({ depA: '2.0.1', depB: '2.0.1', depLib: '2.0.1' })
+    })
   }
 }

